parlai/mturk/tasks/acute_eval/run.py

```diff
@@ -112,341 +102,379 @@ def add_args(from_argv=False):
+        for t in tasks_chosen:
+            worker_data['conversations_seen'].extend(
+                self._get_dialogue_ids(self.desired_tasks[t])
+            )
+            task_data.extend(self.desired_tasks[t])
+        return task_data
 
-        # Soft-block all chosen workers
-        if not opt['is_sandbox'] and opt['softblock_list_path'] is not None:
+    def requeue_task_data(self, worker_id: int, task_data: List[Dict[str, Any]]):
+        """
+        Return task to task_queue.
+
+        If the task is an onboarding task, indicate that the worker has
+        another onboarding task to do.
+
+        :param worker_id:
+            worker id of worker who is returning task
+
+        :param task_data:
+            list of unfinished tasks to return to the queue.
+        """
+        worker_data = self._get_worker_data(worker_id)
+        for subtask_data in task_data:
+            if subtask_data['task_specs'].get('is_onboarding', False):
+                worker_data['onboarding_todo'].append(subtask_data['pair_id'])
+            else:
+                self.task_queue.put(subtask_data)
+                try:
+                    worker_data['tasks_completed'].remove(subtask_data['pair_id'])
+                    for d_id in self._get_dialogue_ids(subtask_data):
+                        worker_data['conversations_seen'].remove(d_id)
+                except ValueError:
+                    print("WARNING: couldn't remove task from worker's history")
+
+    def get_onboarding_tasks(self, worker_id: int) -> List[Dict[str, Any]]:
+        """
+        Get next onboarding task for given worker.
+
+        :param worker_id:
+            worker id
+
+        :return:
+            A list of onboarding tasks for the worker
+        """
+        if len(self.onboarding_tasks) == 0:
+            return []
+
+        worker_data = self._get_worker_data(worker_id)
+        onboarding_todo = worker_data['onboarding_todo']
+        if not onboarding_todo:
+            # worker has completed all required onboarding tasks
+            return []
+        # get onboarding tasks for workers needing them
+        num_tasks_to_return = min(len(onboarding_todo), self.opt['subtasks_per_hit'])
+        onboarding_tasks_chosen = onboarding_todo[:num_tasks_to_return]
+        worker_data['onboarding_todo'] = onboarding_todo[num_tasks_to_return:]
+        return [self.onboarding_tasks[t_id] for t_id in onboarding_tasks_chosen]
+
+    def check_and_update_worker_approval(
+        self, worker_id: int, save_data: Dict[str, Any]
+    ):
+        """
+        Soft block workers who fail onboarding tasks, keep track of their status.
+
+        :param worker_id:
+            worker id
+
+        :param save_data:
+            data from the worker's completed tasks
+        """
+        all_task_data = save_data['worker_data'][worker_id]['task_data']
+        response_data = save_data['worker_data'][worker_id]['response']['task_data']
```

Comment:
what is `response_data` vs. `task_data`?

Reply:
task data is what was served to the client, response data is what was sent back